content/courses/ada-idioms/chapters/inheritance_idioms.rst

```diff
@@ -0,0 +1,577 @@
+completion in the package private part, using a type extension. The steps are
+the same two for any private type: a declaration in the package visible part,
+with a completion in the package private part. The difference is the client
+visibility to the parent type.
+
+
+Solution
+--------
+
+There are two *solutions* in this entry, one for each of the two inheritance
+idioms under discussion. First, we will specify our building block choices,
+then we will show the two idiom expressions in separate subsections.
+
+- We will use tagged types for the sake of the full OOP capabilities. That is
+  the most common choice when inheritance is involved. The static form of
+  inheritance has cases in which it is useful, however those cases are very
+  narrow in applicability.
+
+- We will assume that both the parent type and the child type are both private
+  types, i.e., abstract data types, because that is the best practice. See the
+  :ref:`Abstract Data Type idiom <Ada_Idioms_Abstract_Data_Types>` for
+  justification and details.
+
+- For the most general capabilities, we will assume that the parent type is
+  visibly tagged.
+
+- We are going to declare the child type in a distinct, dedicated package,
+  following the :ref:`ADT idiom <Ada_Idioms_Abstract_Data_Types>`. This package
+  may or may not be a hierarchical child of the parent package. The solution
+  approach does not require a child package's special compile-time visibility,
+  although a child package is often necessary for the sake of that visibility.
+
+- Whether the child type is visibly derived will vary with the
+  :ref:`inheritance idiom <Ada_Idioms_Implementation_Inheritance>` solution.
+
+To avoid unnecessary code duplication, the two idiom solutions use the same
+parent type, declared as a simple tagged private type. The parent type could
+itself be derived from some other tagged type, but that changes nothing
+conceptually significant. The parent type is declared in package :ada:`P`, as
+follows:
+
+.. code-block:: ada
+
+    package P is
+       type Parent is tagged private;  -- visibly tagged
+       --  primitive operations with type Parent as the
+       --  controlling formal parameter
+    private
+       type Parent is tagged record ... end record;
+    end P;
+
+
+.. _Ada_Idioms_Subtype_Inheritance:
+
+Subtype Inheritance
+~~~~~~~~~~~~~~~~~~~
+
+Recall that Subtype Inheritance requires clients to have compile-time
+visibility to the *is-a* relationship between the child and parent types. We
+can satisfy that requirement if we make the child visibly derived from the
+parent. Hence we declare the private type as a private extension in the visible
+part of the package:
+
+.. code-block:: ada
+
+    with P; use P;
+    package Q is
+      type Child is new Parent with private;
+      --  implicit, inherited primitive Parent operations declared here,
+      --  now for type Child additional primitives for Child explicitly
```

Comment:
"primitives for Child explicitly" should start a new comment on its own line

Reply:
Ok, I've just added a newline before the word "additional".